Update ComputerUnix.cs to fix Restart- and Stop-Computer on Linux/Mac

mark-wilson-spc opened the pull request:
Fix `Restart-Computer` on Linux and `Stop-Computer` on MacOS by modifying `RunCommand` to respect `command` and `args` inputs.

# PR Summary

Fix `Restart-Computer` on Linux systems and `Stop-Computer` on MacOS systems.  

See issues [#14684](https://github.com/PowerShell/PowerShell/issues/14684) and [#19738](https://github.com/PowerShell/PowerShell/issues/19738)

Made minimal changes necessary to correct behavior.

I am new at contributing to projects and have not done a build or testing.  Perhaps someone in the Community or on the PowerShell team can validate functionality.

## PR Context

`Restart-Computer` on Linux has been shutting down systems instead of restarting them.. 
`Stop-Computer` on MacOS has been incorrectly throwing errors due to missing arguments.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

mark-wilson-spc committed:
Update ComputerUnix.cs to fix Restart- and Stop-Computer on Linux and Mac

Fix RunCommand to respect 'command' and 'args' inputs.

mark-wilson-spc committed:
Change RunCommand args default value to constant

Can't use string.empty as a default parameter value.  Changed to "".

iSazonov reviewed (commented):
@mklement0 Could you please check the build on Mac and Linux?

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs:
```diff
@@ -153,13 +153,13 @@ protected override void StopProcessing()
         /// Run a command.
         /// </summary>
-        protected void RunCommand(String command, String args) {
+        protected void RunCommand(String command = "/sbin/shutdown", String args = "") {
```
I suggest don't complicate the PR and only fix the code with right `shutdown` for Linux and Mac. Then open new issue to enhance the cmdlet to support `Force` (and grace shutdown) on Unix.

mklement0 replied:
Good point, @iSazonov - I've opened:
* #19791

mark-wilson-spc replied:
@iSazonov defaults have been removed.  I had put them in there just because those were the previous hardcoded values.  Changed `String `to `string `per your request.

mklement0 replied:
Thanks, @mark-wilson-spc; note that changing 

```c#
            if (Platform.IsMacOS)
            {
                args = "now";
            }
```

to

```c#
            if (Platform.IsMacOS)
            {
                args = "-h now";
            }
```

is still necessary.

mark-wilson-spc replied:
Mac `Stop-Computer` args have been changed from `now` to `-h now` .

mark-wilson-spc replied:
@mklement0 changing the args caused the unelevated pester tests to fail.

mklement0 replied:
@mark-wilson-spc,  it is a fundamental requirement on macOS to run `/sbin/shutdown` via `sudo`. In other words: the test is flawed.

mklement0 replied:
@mark-wilson-spc, glancing at the tests, I would have expected them to fail previously as well, given that unelevated invocation with _any_ arguments - including none, as currently - causes `/sbin/shutdown` to fail categorically on macOS, printing `shutdown: NOT super-user` to stderr and terminating with exit code `1`.

mklement0 replied:
@iSazonov, not via `/sbin/shutdown`, specifically, but they _can_ via the GUI and via the AppleScript commands proposed for _graceful_ shutdown / restart proposed in #19791. 

That is, in the latter two cases _no_ administrator role / `sudo` calls are required.

iSazonov replied:
So do we need to run the test with elevated rights?

mark-wilson-spc replied:
Are we open to narrowing this to an atomic change implementing the restart fix for Linux distros, and then moving the outstanding MacOS issues to @mklement0 's [#19791](https://github.com/PowerShell/PowerShell/issues/19791)?

mklement0 replied:
@iSazonov, I haven't dug too deep, but it looks like the `RunCommand()` call is by design never reached  when the test hooks are in place, which indeed makes it mysterious that changing the arguments should make any difference.

It looks like the existing tests are virtually useless on Unix.
To implement meaningful tests (that do not actually _invoke_ `/sbin/shutdown`) would require testing for being `root` and failing if not - both on macOS and Linux. (If the graceful methods are implemented for macOS, that would have to be taken into account.)

Given the lack of actual invocation, the syntax of the arguments can only be verified by comparing the dynamically constructed command line to what it _should be_, as determined via the `man` pages and as verified by interactive invocation (it's safe to assume that this syntax will never change).

mklement0 replied:
@mark-wilson-spc, note that the _restart_ fix would work for macOS too, but _shutdown_ would still be broken, if `"now"` isn't changed to `"-h now"`.

My personal preference is to include the latter change, and then deal with the following aspects separately:
* graceful shutdown/reboot on macOS: #19791
* inability to programmatically detect failure: #19802

Given my discussion of the tests above: Is it possible that the macOS test failure you saw was _transient_, unrelated to your changes? If so, including the `"-h now"` too shouldn't be a problem.

mark-wilson-spc replied:
Thank you for the clarification, @mklement0 .  My thought was that we could go ahead and push out the working fix (`Restart-Computer` for Linux and MacOs) immediately since it is ready to go and has a broad benefit, and then address the thornier shutdown issue separately.

As far as the testing goes, I am new at this and am not in a position to speak to the results of the automated testing (especially where MacOS is concerned).  I had assumed all tests were valid and robust, and that it was necessary to pass all tests for merge approval.  I'm not sure who can best speak to your question regarding the possible transient or false nature of a test result.

mklement0 replied:
Thanks, @mark-wilson-spc - I'll have to defer to others too with respect to the tests. 

As for whether to hold off on the macOS `"-h now"` fix to also make `Stop-Computer` work:

* As noted, even without the fix, `Restart-Computer` _will_ start to work on macOS, because `-r now` works on macOS too.

* If we proceed without the shutdown fix for macOS, we'll have a _somewhat_ working `Restart-Computer` there - one that forces  _instant shutdown with potential data loss_ -  and a still broken `Stop-Computer`. 

* If we include the shutdown fix for macOS we'll have _somewhat_ working `Restart-Computer` and `Stop-Computer` cmdlets on macOS, with the caveats noted.

If _either_ approach gets into a stable release _before_ the graceful approach is implemented on macOS (I'm not holding my breath), we'll have a backward-compatibility problem on our hand: users may by then have come to _rely_ on the forceful and therefore _guaranteed_ shutdown/reboot, even without having specified `-Force`.

Therefore, the most conservative approach would be: fix the problem _for Linux only_ for now (which already had a working `Stop-Computer`, but a broken `Restart-Computer`), which would require deliberately setting the arguments to pass to `shutdown` to something _invalid_ for `Restart-Computer` too.

I hope that lays out the options clearly, and I'll leave the decision to others.

iSazonov replied:
In any case we can not really stop/reboot CI system and have to use test mosc/hook. **So we must check the cmdlets manually before merge.**

As for code, I suggest to implement "-h now" for macOs as @mklement0 proposed. As result after the PR we should get the cmdlets working well on Linux and macOS but without (1) grace scenario, (2) error processing - both are for follow PRs.

mark-wilson-spc replied:
I am working on an implementation and should have it ready later today.

mark-wilson-spc committed:
Update RunCommand params per PR input from iSazinov

Remove default parameter assignments and change 'String' to 'string'.

mark-wilson-spc committed:
Update ComputerUnix.cs change MacOs Stop- args to -h now

Change MacOs 'Stop-Computer' args from 'now' to '-h now' per PR directive.

mklement0 commented on src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs:
```diff
@@ -153,13 +153,13 @@ protected override void StopProcessing()
         /// Run a command.
         /// </summary>
-        protected void RunCommand(String command, String args) {
+        protected void RunCommand(string command, string args) {
```
The way `/sbin/shutdown` is invoked makes it impossible to detect failure on the PowerShell side, though that's probably for a separate PR; see

* #19802

mark-wilson-spc committed:
Update ComputerUnix.cs with MacOs and Unix force and noforce args

Added handling for MacOS and Unix Force and NoForce restarts and shutdowns.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs:
```diff
@@ -21,6 +21,16 @@ namespace Microsoft.PowerShell.Commands
+        /// Force the operation to take place if possible.
+        /// </summary>
+        [Parameter]
+        public SwitchParameter Force { get; set; } = false;
```
No need to init to default.
```suggestion
        public SwitchParameter Force { get; set; }
```

mark-wilson-spc replied:
I will remove it, but just an fyi I used Computer.cs as my style template and it is defaulted there.

mark-wilson-spc replied:
Implemented.

mark-wilson-spc replied:
Removing the default assignment `false` on `force` params may have caused validation tests to fail.  Re-added as a test.

iSazonov replied:
No, it doesn't affect the tests.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs:
```diff
@@ -29,6 +39,36 @@ public sealed class RestartComputerCommand : CommandLineCmdletBase
         /// </summary>
         protected override void BeginProcessing()
         {
+            string unixRestartCommand = "/sbin/shutdown";
+            string unixRestartArgs = "-r now";
+            
+            string macOSRestartCommand = "osascript";
+            string macOSRestartArgs = @"-e 'tell application ""System Events"" to restart'";
+
+            string macOSForceRestartCommand = "/sbin/shutdown";
+            string macOSForceRestartArgs = "-r now";            
```
Please use `string const` for all.

mark-wilson-spc replied:
Implemented.

mark-wilson-spc committed:
Update ComputerUnix.cs - add const to string declarations and remove switch default value

Implement requested changes from ISazonov.
Change string to string const
Remove unnecessary default value of false on Force parameter declarations.

mark-wilson-spc committed:
Update ComputerUnix.cs Restore default value FALSE to FORCE params

Removing the assignment seems to have caused validation testing to fail.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs:
```diff
@@ -29,6 +39,36 @@ public sealed class RestartComputerCommand : CommandLineCmdletBase
         /// </summary>
         protected override void BeginProcessing()
         {
+            string const unixRestartCommand = "/sbin/shutdown";
```
Build failed on the wrong syntax.

mark-wilson-spc replied:
Corrected.

mark-wilson-spc committed:
Update ComputerUnix.cs fix const string syntax error

mark-wilson-spc committed:
Update ComputerUnix.cs fix syntax error

iSazonov commented:
@mark-wilson-spc Please sign CLA (see instructions above) if you agree. We can not merge without CLA.

iSazonov commented:
@mklement0 Please review last build.
We should also implement -WhatIf] / -Confirm support. I hope `osascript` doesn't ask user?

mark-wilson-spc commented:
@microsoft-github-policy-service agree company="SierraSpace.com"

JamesWTruher commented on src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs:
```diff
@@ -29,6 +39,36 @@ public sealed class RestartComputerCommand : CommandLineCmdletBase
+                command = unixRestartCommand;
+                args = unixRestartArgs;
+            }
+            
```
now that you've changed the way shutdown is being called (and adding the more gentle osascript approach, I would be happier if we boosted the tests here, at least to have validation between force and not-force on mac

SteveL-MSFT replied:
In CI, a test hook will have to be needed to verify intended behavior

JamesWTruher reviewed (commented):
I like this (I think we could add sleep too - in a separate PR).
I just have 1 question and a concern about more validation.

SteveL-MSFT commented:
WIll see if someone on the team can finish this PR

SteveL-MSFT reviewed (approved):
I'm ok with the changes as there's an ask to continue to improve this in subsequent PRs

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs:
```diff
@@ -29,6 +39,36 @@ public sealed class RestartComputerCommand : CommandLineCmdletBase
+            const string unixRestartArgs = "-r now";
+            
+            const string macOSRestartCommand = "osascript";
+            const string macOSRestartArgs = @"-e 'tell application ""System Events"" to restart'";
```
ShouldProcess might be appropriate.
